lambda/rotator-snowflake/rotate/rotate.go

```diff
@@ -0,0 +1,169 @@
+				if secret == "" {
+					return errors.Wrapf(err, "key %s has an empty secret", key)
+				}
+				err = secretsAPI.PutSecret([]byte(secret), currentScope, key)
+				if err != nil {
+					return errors.Wrapf(err, "Unable to put secret %s in scope %s", secret, scopeName)
+				}
+			}
+
+			return nil
+		}
+	}
+
+	// create scope called currentScope
+	err = databricks.Secrets().CreateSecretScope(currentScope, "")
+	if err != nil {
+		return errors.Wrapf(err, "Unable to create %s scope with %s perms", currentScope, "")
+	}
+	// Allow admins to manage this secret
+	err = databricks.Secrets().PutSecretACL(currentScope, "admins", models.AclPermissionManage)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to make admins control this scope: %s", currentScope)
+	}
+
+	return updateDatabricks(currentScope, secrets, databricks)
+}
+
+func updateSnowflake(user string, db *sql.DB, privKeyBuffer *bytes.Buffer) error {
+	privPemBlock, _ := pem.Decode(privKeyBuffer.Bytes())
+
+	privKeyIface, err := x509.ParsePKCS8PrivateKey(privPemBlock.Bytes)
+	if err != nil {
+		return errors.Wrap(err, "Unable to get pkcs8 private key")
+	}
+
+	privKey, ok := privKeyIface.(*rsa.PrivateKey)
+	if !ok {
+		return errors.New("Unable to get pkcs8 private key")
+	}
+
+	publicKeyBytes, err := x509.MarshalPKIXPublicKey(&privKey.PublicKey)
+	if err != nil {
+		return errors.Wrap(err, "Unable to marshal public key to bytes")
+	}
+
+	publicKeyStr := base64.StdEncoding.EncodeToString(publicKeyBytes)
+	query := fmt.Sprintf(`ALTER USER "%s" SET RSA_PUBLIC_KEY_2 = "%s"`, user, publicKeyStr)
+	_, err = snowflake.ExecNoRows(db, query)
+
+	return err
+}
+
+func Rotate(ctx context.Context) error {
+	snowflakeDB, err := setup.Snowflake()
+	if err != nil {
+		return errors.Wrap(err, "Unable to configure snowflake")
+	}
+
+	databricksConnection, err := setup.Databricks()
+	if err != nil {
+		return errors.Wrap(err, "Unable to configure databricks")
+	}
+
+	users, err := getUsers()
+	if err != nil {
+		return errors.Wrap(err, "Unable to get list of users to rotate")
+	}
+
+	// Collect errors for each user:
+	userErrors := &multierror.Error{}
+
+	for _, user := range users {
+		privKey, err := keypair.GenerateRSAKeypair()
+		if err != nil {
+			return errors.Wrap(err, "Unable to generate RSA keypair")
+		}
+
+		privKeyBuffer, err := keypair.SaveRSAKey(privKey)
+		if err != nil {
+			return errors.Wrap(err, "Unable to save RSA Keypair")
+		}
+
+		err = updateSnowflake(user, snowflakeDB, privKeyBuffer)
+		if err != nil {
+			userErrors = multierror.Append(userErrors, err)
+		}
+
+		snowflakeSecrets, err := buildSnowflakeSecrets(snowflakeDB, user, privKeyBuffer)
+		if err != nil {
+			return errors.Wrap(err, "Cannot generate Snowflake Secrets Map")
+		}
+
+		// Intentionally equating databricks scope and user here
+		err = updateDatabricks(user, snowflakeSecrets, databricksConnection)
+		if err != nil {
+			userErrors = multierror.Append(userErrors, err)
+		}
+	}
```

Comment:
I like the multierror approach here, what do you think about handling all errors this way?

```suggestion
	// Collect errors for each user:
	userErrors := &multierror.Error{}
        processUser := func(user string) error {
		privKey, err := keypair.GenerateRSAKeypair()
		if err != nil {
			return errors.Wrap(err, "Unable to generate RSA keypair")
		}

		privKeyBuffer, err := keypair.SaveRSAKey(privKey)
		if err != nil {
			return errors.Wrap(err, "Unable to save RSA Keypair")
		}

		err = updateSnowflake(user, snowflakeDB, privKeyBuffer)
		if err != nil {
			return //todo
		}

		snowflakeSecrets, err := buildSnowflakeSecrets(snowflakeDB, user, privKeyBuffer)
		if err != nil {
			return errors.Wrap(err, "Cannot generate Snowflake Secrets Map")
		}

		// Intentionally equating databricks scope and user here
		return updateDatabricks(user, snowflakeSecrets, databricksConnection)
	}
	for _, user := range users {
	   err := processUser(user)
	   if err != nil {
	      userErrors = multierror.Append(userErrors, err)
	   }
	   return userErrors.ErrorOrNil()
	}
```